oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/PersistenceTimer.java

```diff
@@ -63,98 +70,149 @@ public void start(ModuleManager moduleManager, CoreModuleConfig moduleConfig) {
-                persistenceWorkers.addAll(TopNStreamProcessor.getInstance().getPersistentWorkers());
-                persistenceWorkers.addAll(MetricsStreamProcessor.getInstance().getPersistentWorkers());
-
-                persistenceWorkers.forEach(worker -> {
-                    if (log.isDebugEnabled()) {
-                        log.debug("extract {} worker data and save", worker.getClass().getName());
+            List<PersistenceWorker> persistenceWorkers = new ArrayList<>();
+            persistenceWorkers.addAll(TopNStreamProcessor.getInstance().getPersistentWorkers());
+            persistenceWorkers.addAll(MetricsStreamProcessor.getInstance().getPersistentWorkers());
+            CountDownLatch countDownLatch = new CountDownLatch(MetricsStreamProcessor.getInstance().getPersistentWorkers().size());
+
+            persistenceWorkers.forEach(worker -> {
+                prepareExecutorService.submit(() -> {
+                    HistogramMetrics.Timer timer = prepareLatency.createTimer();
+                    try {
+                        if (log.isDebugEnabled()) {
+                            log.debug("extract {} worker data and save", worker.getClass().getName());
+                        }
+                        List<PrepareRequest> innerPrepareRequests = new ArrayList<>(5000);
+                        worker.buildBatchRequests(innerPrepareRequests);
+                        synchronized (prepareRequests) {
+                            prepareRequests.addAll(innerPrepareRequests);
+                            if (prepareRequests.size() >= maxSyncoperationNum) {
+                                prepareRequests.notify();
+                            }
+                        }
+                        worker.endOfRound(System.currentTimeMillis() - lastTime);
+                    } finally {
+                        timer.finish();
+                        countDownLatch.countDown();
                     }
-
-                    worker.buildBatchRequests(prepareRequests);
-
-                    worker.endOfRound(System.currentTimeMillis() - lastTime);
                 });
+            });
+
+            Future<?> batchFuture = batchExecutorService.submit(() -> {
+                List<Future<?>> results = new ArrayList<>();
+                while (true) {
+                    synchronized (prepareRequests) {
+                        if (prepareDone && CollectionUtils.isEmpty(prepareRequests)) {
+                            break;
+                        }
+                    }
 
-                if (debug) {
-                    log.info("build batch persistence duration: {} ms", System.currentTimeMillis() - startTime);
-                }
-            } finally {
-                timer.finish();
-            }
+                    synchronized (prepareRequests) {
+                        while (this.prepareRequests.size() < maxSyncoperationNum && !prepareDone) {
```

Comment:
If there is a resource available, why don't run persistence when you have `prepareRequests`?

Reply:
If we execute a very small number at a time, it may cause the previous threads to perform a small number of writes and each thread waits for 10 seconds in the end (wait for the refresh. ). I think the time to fill the prepareRequests faster than wait 10 sec.

Reply:
It may execute a small set in the first time, but wouldn't be larger in the next iteration? Because you have very limited threads. If you just wait, and nothing happens. From the whole picture, I think we cost more time?